src/system/SystemObject.h

```diff
@@ -184,7 +184,11 @@ class ObjectPool
 private:
     friend class TestObject;
 
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+    std::vector<std::unique_ptr<T>> mObjects;
```

Comment:
Following up to my comment, https://github.com/project-chip/connectedhomeip/pull/8299#issuecomment-879536059, I don't think this is going to help.

I verified there is in fact lots of code in InetLayer and SystemLayer that does iterate through the sPools.  So yes, we need to have the ability to inspect the pool contents.

But using a std::vector won't fix the problem that large systems are going to have.  Yes, this moves the decision for pool size from compile time to runtime.  But you still need to decide on a pool size when the program starts.  That's not good enough.  Actually, I don't think it's useful improvement at all.  We need to be able to allocate while the program is running.

This is what malloc is for.  The way this could be used here is with a linked list.  Malloc new nodes and add them to the linked list; free nodes when no longer needed and remove them from the linked list.  Various stack layers can still iterate through the sPools, but would need to be #ifdefed to traverse the linked list for `CHIP_SYSTEM_CONFIG_POOL_USE_HEAP` builds.

A quick search shows 10 places where sPool iterating code would need to be changed to traverse a linked list.  The compiler can find the others.

```
find . -name "*.cpp" -exec grep "for (size_t i = 0; i < .*::sPool" {} +
./system/SystemLayer.cpp:    for (size_t i = 0; i < Timer::sPool.Size(); ++i)
./system/SystemLayer.cpp:    for (size_t i = 0; i < Timer::sPool.Size(); ++i)
./system/SystemLayer.cpp:    for (size_t i = 0; i < Timer::sPool.Size(); i++)
./system/SystemLayer.cpp:    for (size_t i = 0; i < Timer::sPool.Size(); i++)
./inet/InetLayer.cpp:        for (size_t i = 0; i < DNSResolver::sPool.Size(); i++)
./inet/InetLayer.cpp:        for (size_t i = 0; i < RawEndPoint::sPool.Size(); i++)
./inet/InetLayer.cpp:        for (size_t i = 0; i < TCPEndPoint::sPool.Size(); i++)
./inet/InetLayer.cpp:        for (size_t i = 0; i < UDPEndPoint::sPool.Size(); i++)
./inet/InetLayer.cpp:    for (size_t i = 0; i < TCPEndPoint::sPool.Size(); i++)
./inet/InetLayer.cpp:    for (size_t i = 0; i < DNSResolver::sPool.Size(); i++)
```

Reply:
I should note that we could move to just using new/delete, with overrides of those operators on the relevant classes that do the following:
1. When using a pool, allocate from the pool, returning null when pool is full.
2. When not using a pool, heap-allocate via global new and then add to list (and delete removes from list).

This probably is probably simplest with a compile-time flag (for the extra members to keep track of list membership), though we could make it runtime-selectable too, with some list-member helper objects...

In any case, if we do this, the user-facing API for allocation/deallocation would be the same in either case: just use new/delete.

Reply:
We don't need to decide the pool size up front, the vector mObjects is empty at the beginning. The object is allocated and wrapped with a unique pointer in runtime based on the actual need.

I don't see much difference between the method you described and current implementation with this PR, except the object is allocated with 'new' instead of 'malloc' and managed by vector instead of linked list

Reply:
I think the main objection is the fact that there's still a compile-time-decided cap on the number of objects that can be allocated from the "pool".

> except the objects are allocated with 'new' instead of 'malloc'

That's just ergonomics, not functionality.

> and managed by vector instead of linked list

The difference is what happens if I allocate 10,000 objects and then free them.  With the linked list approach we free all the relevant memory.  With the vector approach in this PR we have a vector of 10,000 nulled-out unique_ptr sitting around.

Reply:
> I think the main objection is the fact that there's still a compile-time-decided cap on the number of objects that can be allocated from the "pool".
> 
> > except the objects are allocated with 'new' instead of 'malloc'
> 
> That's just ergonomics, not functionality.
> 
> > and managed by vector instead of linked list
> 
> The difference is what happens if I allocate 10,000 objects and then free them. With the linked list approach we free all the relevant memory. With the vector approach in this PR we have a vector of 10,000 unique_ptr sitting around each one pointing to a still-live object.

When I proposed this approach I fully expected we would be able to erase the elements of the vector. But apparently that's not the case.

But if we can remove the item from a link list, we can erase it from the vector. So I don't think think the link list idea is really an improvement at all.

Reply:
I wasn't proposing that we use a linked-list to track elements of a vector.  I was proposing that we not try to allocate anything (whether they be references or objects themselves) in contiguous memory at all, and instead allocate each element individually.  Then we can track all of the allocated elements in a linked list so that the various layers can act on them.  This could be easily realized by adding previous and next pointers to `Object`.

This alleviates the need to have to decide ahead of time the maximum number of elements that should be supported, which I think is the crux of the problem on large systems.

Reply:
So maybe I have misunderstood this PR.  If we are storing references, then I suppose we can dynamically increase the sizes of the pools because this means we can relocate the underlying memory for the vectors.

But I guess I still don't see a compelling reason to use contiguous memory.  There's no need for it and this forces O(N) traversals for allocations and other operations.  And it means we will often have times when indices of the pools are sitting empty for no good reason.  Certainly, the linear pools aren't somehow helping fragmentation (which I would argue isn't an issue for the target use cases anyway) because we're still allocating the actual pool objects individually.

A linked list just feels like the right data structure for this.

Reply:
Also, if we do care about fragmentation, this is in theory actually even worse because we only ever allocate objects in arbitrary locations of memory and never free them.

Reply:
I have no real problem with using a link list, it's just not a particularly friendly data structure for the memory subsystem (cache hostile).

I'm honestly not too interested in whether it's std::list<T> vs std::vector<std::unique_ptr<T>>. The real issue is that we apparently can't do either of those things easily, because of how deallocation works. So that's where we should focus.

Reply:
I'm proposing that the allocated objects be the list nodes themselves:

> This could be easily realized by adding previous and next pointers to Object.

This isn't more cache hostile.  The objects are distributed in memory exactly as they would be in the current PR.  But there wouldn't be a second data structure to track them.  This also means that the object can remove itself from the list from within the Release method and, so long as the base Object class destructor is changed to virtual,  the object can also go ahead and delete itself.

Reply:
The object in the pool is allocated individually, they are not in a contiguous memory,  the unique_ptr is, but do we care?

Reply:
Iterating over the vector is definitely not great for scaling to large pools.  The solution I proposed wouldn't have that problem.  Now that you've solved the deallocation problem, this is really the only remaining difference.

Reply:
I would like to give a try if std::list<T> has better permanence over std::vector<std::unique_ptr<T>> on iteration.

Reply:
> I would like to give a try if std::list has better permanence over std::vector<std::unique_ptr> on iteration.

Suspect you'll have an issue converting a T* to a std::list<T>::iterator and we'll be back to a linear search, which will be worse with a list.

But seriously, it doesn't matter.

Reply:
Again, I'm not talking about `std::list<T>`.  Maintaining a separate data structure is counter productive.  The allocated elements can be the list nodes themselves:

```
diff --git a/src/system/SystemObject.cpp b/src/system/SystemObject.cpp
index 3b469362134..9f221aad3f2 100644
--- a/src/system/SystemObject.cpp
+++ b/src/system/SystemObject.cpp
@@ -51,7 +51,9 @@ DLL_EXPORT void Object::Release()
 
     if (oldCount == 1)
     {
-        this->mSystemLayer = nullptr;
+        this->prev->next = this->next;
+        this->next->prev = this->prev;
+        delete this;
         __sync_synchronize();
     }
     else if (oldCount == 0)
diff --git a/src/system/SystemObject.h b/src/system/SystemObject.h
index 4b74cdd7eb8..ca8f48eda87 100644
--- a/src/system/SystemObject.h
+++ b/src/system/SystemObject.h
@@ -97,9 +97,11 @@ protected:
 
 private:
     Object();
-    ~Object();
+    virtual ~Object();
     Object(const Object &) = delete;
     Object & operator=(const Object &) = delete;
+    Object *next;
+    Object *prev;
```

This will obviously scale better.  I'm not insisting we do this.  But it solves an actual problem.

Reply:
Ok, sure, let's do the intrusive list.

Reply:
"intrusive" refers to any container that maintains its state directly inside its elements. So this data structure is an "intrusive list". Contrast std::list which is a non-intrusive list.

Reply:
(i.e., it wasn't a complaint, I'm agreeing with you)

Reply:
Oh, I see.  Got it.

Reply:
I tried to update my PR with linked list and release each element individually, but I hit various stability issues, the main reason is the object was created from ObjectPool:  T * ObjectPool<T, N>::TryCreate(Layer & aLayer). but the release of object can be done by itself something like:

DLL_EXPORT void Object::Release()
{
    unsigned int oldCount = __sync_fetch_and_sub(&this->mRefCount, 1);

    if (oldCount == 1)
    {
        this->mSystemLayer = nullptr;
#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP

        if (this->prev == nullptr)
        {
            *mHeadRef = nullptr;
        }
        else
        {
            this->prev->next = this->next;
            if (this->next)
                this->next->prev = this->prev;
        }
        delete this;
#endif
        __sync_synchronize();
    }
    else if (oldCount == 0)
    {
        abort();
    }
}

I need to keep a reference to the head of linked list in case the head need to be updated. How to keep the create and release are synced with this solution?

Reply:
> I tried to update my PR with linked list and release each element individually, but I hit various stability issues in multi-thread testing, the main reason is the object was created from ObjectPool: T * ObjectPool<T, N>::TryCreate(Layer & aLayer). but the release of object is done by itself something like:

The locking needs do not change.  You would also need a pool mutex, would need a pointer to this in the object, and would need to acquire the lock while disconnecting from the list in the Release method.

> I need to keep a reference to the head of linked list in case the head need to be updated. How to keep the create and release are synced with this solution?

The easiest way to do this is to have a dummy object as a member of each pool that can point to allocated objects.  This is better than a reference to the head object because then the list-disconnection logic inside each object does not need to change for removal of the last list node, and also implicitly updates the pool list pointer even on disconnection (Release) of the last object.

Reply:
Just as a FYI, I should have removed all Timer pool iteration within a week, which leaves only the Inet cases.

Reply:
@msandstedt I have updated the code with double linked list solution, it passed all multi-threading tests with my local tests, lets see if it can pass tsan CI tests

Reply:
Thank you.

I had one additional comment about O(1) insertion at allocation time.

All else looks good to me.

Reply:
That is my original plan to add the node at the head instead of tail, but it randomly failed on CI, I don't have explanation, lets see how it goes this time

Reply:
Failed as expected, I think there is race in the retention test since I get the node from the head TestObject * lGotten = sPool.Get(lLayer, 0); will update the unit test case.

Reply:
The indexed getter is the root of the problem, isn't it?  This no longer makes sense.  In addition to the now O(N^2) behavior of iteration, iteration is actually broken in the sdk because it's performed without the lock held:

```
        for (size_t i = 0; i < RawEndPoint::sPool.Size(); i++)
        {
            RawEndPoint * lEndPoint = RawEndPoint::sPool.Get(*mSystemLayer, i);
            if ((lEndPoint != nullptr) && lEndPoint->IsCreatedByInetLayer(*this))
            {
                lEndPoint->Close();
            }
        }
```

I would suggest removing the indexed getter and replacing it with some type of templated foreach method that can safely execute some type of operation on each allocated object.  External code doesn't need to know about pool indices, which don't really exist now anyway in the heap implementation.

Reply:
The failure only occurs in CI but not locally, the CI show we have seg fault right after complete  'Simple Init Test IPV4' test case.

2021-08-01T23:10:11.5068516Z '#3:','Simple Init Test IPV4     ','PASSED'
2021-08-01T23:10:11.5069012Z [1627859411.037667][11680:11680] CHIP:IN: TransportMgr initialized
2021-08-01T23:10:11.5069517Z CHIP node ready to service events
2021-08-01T23:10:11.5070065Z [1627859411.038774][11680:11680] CHIP:IN: Freeing connection: connection closed by peer
2021-08-01T23:10:11.5070634Z Segmentation fault (core dumped)
2021-08-01T23:10:11.5071071Z INF Test 1/1: [FAIL] TestTCP

Close endpoint does not free the object from the pool, otherwise, we should see the problem easily from local run.

Reply:
The loop logic has to change.  It is a bug to have locking in the pools, but expose indexed getters that imply the indices are stable when they are not.

I suggest fixing this and then revisiting the tests, because those will be affected as well.

Reply:
Updated with ForEachActiveObject, the timer needs to be released during the iteration, I adjusted the logic in ForEachActiveObject to move to next node first before taking operation on the node to overcome this issue. Kevin is working on removing on all Timer pool iteration, then we don't have any case which modify the pool list during the iteration. Hope that is the root cause of CI failure.

Reply:
> Updated with ForEachActiveObject, the timer needs to be released during the iteration, I adjusted the logic in ForEachActiveObject to move to next node first before taking operation on the node to overcome this issue. Kevin is working on removing on all Timer pool iteration, then we don't have any case which modify the pool list during the iteration. Hope that is the root cause of CI failure.

Thanks.  For the most part, this all looks like it will work.  However, `ForEachActiveObjectInner` also needs to acquire the lock.  This does mean that we either cannot allow the lambdas to release or allocate objects, or we must use a recursive lock.  I can't think of any other solutions.

It sounds like we're going to use the former solution and disallow pool list modification during iteration.

Reply:
We still have two places which modify list during iteration, Kevin has the change to remove them. Then we can add lock here and disable lambdas to release or allocate objects during iteration from now on.

Reply:
All open issues have been closed, please take a final look.